java.lang.SecurityException: `Permission Denial`

```
FATAL EXCEPTION: main
Process: io.treehouses.remote, PID: 26395
java.lang.SecurityException: Permission Denial: starting Intent { act=android.bluetooth.adapter.action.REQUEST_ENABLE cmp=com.android.settings/.bluetooth.RequestPermissionActivity } from ProcessRecord{e433290 26395:io.treehouses.remote/u0a276} (pid=26395, uid=10276) requires android.permission.BLUETOOTH_CONNECT
	at android.os.Parcel.createExceptionOrNull(Parcel.java:3023)
	at android.os.Parcel.createException(Parcel.java:3007)
	at android.os.Parcel.readException(Parcel.java:2990)
	at android.os.Parcel.readException(Parcel.java:2932)
	at android.app.IActivityTaskManager$Stub$Proxy.startActivity(IActivityTaskManager.java:2791)
	at android.app.Instrumentation.execStartActivity(Instrumentation.java:1801)
	at android.app.Activity.startActivityForResult(Activity.java:5596)
	at androidx.activity.ComponentActivity.startActivityForResult(ComponentActivity.java:568)
	at androidx.core.app.ActivityCompat.startActivityForResult(ActivityCompat.java:234)
	at androidx.activity.ComponentActivity$2.onLaunch(ComponentActivity.java:215)
	at androidx.activity.result.ActivityResultRegistry$4.launch(ActivityResultRegistry.java:198)
	at androidx.activity.result.ActivityResultLauncher.launch(ActivityResultLauncher.java:42)
	at androidx.fragment.app.FragmentManager.launchStartActivityForResult(FragmentManager.java:2981)
	at androidx.fragment.app.Fragment.startActivityForResult(Fragment.java:1409)
	at androidx.fragment.app.Fragment.startActivityForResult(Fragment.java:1383)
	at io.treehouses.remote.ui.home.HomeFragment.connectRpiListener$lambda-11(HomeFragment.kt:202)
	at io.treehouses.remote.ui.home.HomeFragment.$r8$lambda$npHI9oJjU-uAF9Iv1hMPTScMn68(Unknown Source:0)
	at io.treehouses.remote.ui.home.HomeFragment$$ExternalSyntheticLambda3.onClick(Unknown Source:2)
	at android.view.View.performClick(View.java:7892)
	at android.widget.TextView.performClick(TextView.java:16220)
	at android.view.View.performClickInternal(View.java:7869)
	at android.view.View.-$$Nest$mperformClickInternal(Unknown Source:0)
	at android.view.View$PerformClick.run(View.java:30880)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:226)
	at android.os.Looper.loop(Looper.java:313)
	at android.app.ActivityThread.main(ActivityThread.java:8757)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067)
Caused by: android.os.RemoteException: Remote stack trace:
	at com.android.server.wm.ActivityTaskSupervisor.checkStartAnyActivityPermission(ActivityTaskSupervisor.java:1305)
	at com.android.server.wm.ActivityStarter.executeRequest(ActivityStarter.java:1387)
	at com.android.server.wm.ActivityStarter.execute(ActivityStarter.java:1025)
	at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1942)
	at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1815)
```

--- a/app/src/main/AndroidManifest.xml
+++ b/app/src/main/AndroidManifest.xml
@@ -4,6 +4,8 @@
     <!-- Min/target SDK versions (<uses-sdk>) managed by build.gradle -->
     <uses-permission android:name="android.permission.BLUETOOTH" />
     <uses-permission android:name="android.permission.BLUETOOTH_ADMIN" />
+    <uses-permission android:name="android.permission.BLUETOOTH_CONNECT" />
+    <uses-permission android:name="android.permission.BLUETOOTH_SCAN" />
     <uses-permission android:name="android.permission.INTERNET" />
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
     <uses-permission android:name="android.permission.ACCESS_WIFI_STATE" />

--- a/app/src/main/kotlin/io/treehouses/remote/bases/BaseBluetoothChatService.kt
+++ b/app/src/main/kotlin/io/treehouses/remote/bases/BaseBluetoothChatService.kt
@@ -2,6 +2,7 @@ package io.treehouses.remote.bases
 
 import android.app.Notification
 import android.app.PendingIntent
+import android.app.PendingIntent.FLAG_IMMUTABLE
 import android.app.Service
 import android.bluetooth.BluetoothAdapter
 import android.bluetooth.BluetoothDevice
@@ -54,9 +55,7 @@ open class BaseBluetoothChatService @JvmOverloads constructor(handler: Handler? 
     }
 
     open fun start() {}
-
-
-
+    
     /**
      * Update UI title according to the current state of the chat connection
      */
@@ -82,10 +81,10 @@ open class BaseBluetoothChatService @JvmOverloads constructor(handler: Handler? 
 
     protected fun startNotification() {
         val disconnectIntent = Intent(DISCONNECT_ACTION)
-        val disconnectPendingIntent: PendingIntent = PendingIntent.getBroadcast(this, 0, disconnectIntent, 0)
+        val disconnectPendingIntent: PendingIntent = PendingIntent.getBroadcast(this, 0, disconnectIntent, FLAG_IMMUTABLE)
 
         val onClickIntent = Intent(this, InitialActivity::class.java)
-        val pendingClickIntent = PendingIntent.getActivity(this, 0, onClickIntent, PendingIntent.FLAG_UPDATE_CURRENT)
+        val pendingClickIntent = PendingIntent.getActivity(this, 0, onClickIntent, FLAG_IMMUTABLE)
 
         val notificationBuilder: NotificationCompat.Builder = NotificationCompat.Builder(this, getString(R.string.bt_notification_ID))
         val notification: Notification = notificationBuilder.setOngoing(true)

--- a/app/src/main/kotlin/io/treehouses/remote/bases/BaseBluetoothChatService.kt
+++ b/app/src/main/kotlin/io/treehouses/remote/bases/BaseBluetoothChatService.kt
@@ -55,7 +55,7 @@ open class BaseBluetoothChatService @JvmOverloads constructor(handler: Handler? 
     }
 
     open fun start() {}
-    
+
     /**
      * Update UI title according to the current state of the chat connection
      */

--- a/app/src/main/kotlin/io/treehouses/remote/bases/PermissionActivity.kt
+++ b/app/src/main/kotlin/io/treehouses/remote/bases/PermissionActivity.kt
@@ -39,11 +39,21 @@ abstract class PermissionActivity : AppCompatActivity() {
     }
 
     fun requestPermission() {
-        if (!checkPermission(Manifest.permission.ACCESS_FINE_LOCATION) || !checkPermission(Manifest.permission.ACCESS_COARSE_LOCATION) || !checkPermission(Manifest.permission.CHANGE_WIFI_STATE)) {
+        if (!checkPermission(Manifest.permission.ACCESS_FINE_LOCATION) ||
+            !checkPermission(Manifest.permission.ACCESS_COARSE_LOCATION) ||
+            !checkPermission(Manifest.permission.CHANGE_WIFI_STATE) ||
+            !checkPermission(Manifest.permission.BLUETOOTH) ||
+            !checkPermission(Manifest.permission.BLUETOOTH_ADMIN) ||
+            !checkPermission(Manifest.permission.BLUETOOTH_CONNECT) ||
+            !checkPermission(Manifest.permission.BLUETOOTH_CONNECT)) {
             ActivityCompat.requestPermissions(this, arrayOf(
-                    Manifest.permission.CHANGE_WIFI_STATE,
-                    Manifest.permission.ACCESS_FINE_LOCATION,
-                    Manifest.permission.ACCESS_COARSE_LOCATION
+                Manifest.permission.CHANGE_WIFI_STATE,
+                Manifest.permission.ACCESS_FINE_LOCATION,
+                Manifest.permission.ACCESS_COARSE_LOCATION,
+                Manifest.permission.BLUETOOTH,
+                Manifest.permission.BLUETOOTH_ADMIN,
+                Manifest.permission.BLUETOOTH_CONNECT,
+                Manifest.permission.BLUETOOTH_SCAN,
             ), PERMISSION_REQUEST_WIFI)
         } else {
             statusCheck()

